### PR TITLE
PaddlePaddle Hackathon 54 提交

### PR DESCRIPTION
**Pull request by JunnYu**
Task: #1076

- 权重文件等百度云上传，链接：https://pan.baidu.com/s/1sonj-M_qVTkvRzYdM781Kg 提取码：79s9。
- huggingface的ElectraGeneratorPredictions中nn.LayerNorm没有设置eps=1e-12,paddlenlp中设置了1e-12。
- hfl-chinese-legal-electra-small-generator权重官方的config配置存在错误，他错误的使用了discriminator的配置，因此我人工修正了一下。
- 添加compare.py比较转换后的模型预测结果。（该文件所处位置等审核通过后再修改）。
- 添加convert.py转换脚本。（该文件所处位置等审核通过后再修改）。
- 添加ElectraForMaskedLM， ElectraForQuestionAnswering和ElectraForMultipleChoice这三个类。
- 支持2d的attention mask。
- 添加单元测试代码。TestElectraForMaskedLM，TestElectraForQuestionAnswering和TestElectraForMultipleChoice。
- windows条件下，paddle.int可能是int32，显示转换成int64。
- 添加好多注释，ElectraModel设置默认初始参数。

### CODE REVIEW

**Comment by yingyibiao** [comment]
community/junnyu目录下面的各个文件夹里面只需要 ``README.md`` 和 ``files.json`` 这两个文件，其他在百度网盘上传即可。

**Review by yingyibiao — COMMENTED** [review]
如上

**Comment by JunnYu** [comment]
> community/junnyu目录下面的各个文件夹里面只需要 `README.md` 和 `files.json` 这两个文件，其他在百度网盘上传即可。

修改完成，发现在我提交这个PR之后，`paddle develop`中竟然更新了部分electra模型的文档，然后我这边修复了当中的部分错误。

**Comment by yingyibiao** [comment]
> Task: #1076
> 
> * 权重文件等百度云上传，链接：https://pan.baidu.com/s/1sonj-M_qVTkvRzYdM781Kg 提取码：79s9。
> * huggingface的ElectraGeneratorPredictions中nn.LayerNorm没有设置eps=1e-12,paddlenlp中设置了1e-12。
> * hfl-chinese-legal-electra-small-generator权重官方的config配置存在错误，他错误的使用了discriminator的配置，因此我人工修正了一下。
> * 添加compare.py比较转换后的模型预测结果。（该文件所处位置等审核通过后再修改）。
> * 添加convert.py转换脚本。（该文件所处位置等审核通过后再修改）。
> * 添加ElectraForMaskedLM， ElectraForQuestionAnswering和ElectraForMultipleChoice这三个类。
> * 支持2d的attention mask。
> * 添加单元测试代码。TestElectraForMaskedLM，TestElectraForQuestionAnswering和TestElectraForMultipleChoice。
> * windows条件下，paddle.int可能是int32，显示转换成int64。
> * 添加好多注释，ElectraModel设置默认初始参数。

权重相关文件已上传bos

**Comment by yingyibiao** [comment]
参照 https://github.com/PaddlePaddle/PaddleNLP/pull/1085 的review意见修改类似问题～

File: paddlenlp/transformers/electra/modeling.py [hunk 1]
```diff
@@ -449,6 +447,11 @@ def forward(self,
                 (input_ids == self.pad_token_id
                  ).astype(paddle.get_default_dtype()) * -1e9,
                 axis=[1, 2])
+        else:
+            if attention_mask.ndim == 2:
+                attention_mask = attention_mask.unsqueeze(axis=[1, 2])
+                attention_mask = (1.0 - attention_mask
+                                  ).astype(paddle.get_default_dtype()) * -1e9
```
Comment (yingyibiao) [inline paddlenlp/transformers/electra/modeling.py]: 这里应该也是冗余的判断，可以删除了

File: paddlenlp/transformers/electra/modeling.py [hunk 2]
```diff
@@ -550,7 +553,7 @@ def __init__(self, electra):
                 self.electra.config["embedding_size"],
                 self.electra.config["vocab_size"])
         else:
-            self.generator_lm_head_bias = paddle.fluid.layers.create_parameter(
+            self.generator_lm_head_bias = paddle.create_parameter(
```
Comment (yingyibiao) [inline paddlenlp/transformers/electra/modeling.py]: paddle.create_parameter -> self.create_parameter

File: paddlenlp/transformers/electra/modeling.py [hunk 3]
```diff
@@ -1124,3 +1275,83 @@ def forward(self, generator_prediction_scores,
...
+            position_ids=position_ids,
+            attention_mask=attention_mask)
+        logits = self.classifier(sequence_output)
+        logits = paddle.transpose(logits, perm=[2, 0, 1])
+        start_logits, end_logits = paddle.unstack(x=logits, axis=0)
+
+        return start_logits, end_logits
+
+
+ElectraForMaskedLM = ElectraGenerator
```
Comment (yingyibiao) [inline paddlenlp/transformers/electra/modeling.py]: 这行代码代表什么意思？

File: paddlenlp/transformers/electra/modeling.py [hunk 4]
```diff
@@ -1038,6 +1049,146 @@ def forward(self,
...
+        self.activation = nn.Tanh()
+        self.pool_act = pool_act
+
+    def forward(self, hidden_states):
+        # We "pool" the model by simply taking the hidden state corresponding
+        # to the first token.
+        first_token_tensor = hidden_states[:, 0]
+        pooled_output = self.dense(first_token_tensor)
+        if self.pool_act == "tanh":
+            pooled_output = self.activation(pooled_output)
```
Comment (yingyibiao) [inline paddlenlp/transformers/electra/modeling.py]: 我看hf的summary_activation是gelu，这里为啥是tanh？

File: community/junnyu/hfl-chinese-legal-electra-small-generator/README.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+model = ElectraGenerator.from_pretrained(path)
+tokenizer = ElectraTokenizer.from_pretrained(path)
+model.eval()
+
+text = "欢迎使用paddlenlp！"
+inputs = {
+    k: paddle.to_tensor(
+        v, dtype="int64").unsqueeze(0)
+    for k, v in tokenizer(text).items()
+}
```
Comment (yingyibiao) [inline community/junnyu/hfl-chinese-legal-electra-small-generator/README.md]: 这里的示例需要更加贴近具体权重特点

File: community/junnyu/hfl-chinese-legal-electra-small-generator/README.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+谷歌和斯坦福大学发布了一种名为 ELECTRA 的新预训练模型，与 BERT 及其变体相比，该模型具有非常紧凑的模型尺寸和相对具有竞争力的性能。 为进一步加快中文预训练模型的研究，HIT与科大讯飞联合实验室（HFL）发布了基于ELECTRA官方代码的中文ELECTRA模型。 与 BERT 及其变体相比，ELECTRA-small 只需 1/10 的参数就可以在几个 NLP 任务上达到相似甚至更高的分数。
+这个项目依赖于官方ELECTRA代码: https://github.com/google-research/electra
+该模型是small版本的generator，并且该模型专为法律领域而设计。
+
+# 使用示例
+
+```python
+from paddlenlp.transformers import ElectraGenerator,ElectraTokenizer
+
+path = "hfl-chinese-legal-electra-small-generator"
```
Comment (yingyibiao) [inline community/junnyu/hfl-chinese-legal-electra-small-generator/README.md]: 这里的path需要修改

File: community/junnyu/hfl-chinese-electra-180g-small-ex-discriminator/README.md [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+谷歌和斯坦福大学发布了一种名为 ELECTRA 的新预训练模型，与 BERT 及其变体相比，该模型具有非常紧凑的模型尺寸和相对具有竞争力的性能。 为进一步加快中文预训练模型的研究，HIT与科大讯飞联合实验室（HFL）发布了基于ELECTRA官方代码的中文ELECTRA模型。 与 BERT 及其变体相比，ELECTRA-small 只需 1/10 的参数就可以在几个 NLP 任务上达到相似甚至更高的分数。
+这个项目依赖于官方ELECTRA代码: https://github.com/google-research/electra
+该模型是small版本的discriminator，并且在180G的中文数据上进行训练。
+
+# 使用示例
+
+```python
+from paddlenlp.transformers import ElectraDiscriminator,ElectraTokenizer
+
+path = "hfl-chinese-electra-180g-small-ex-discriminator"
```
Comment (yingyibiao) [inline community/junnyu/hfl-chinese-electra-180g-small-ex-discriminator/README.md]: path需要进行修改

File: community/junnyu/hfl-chinese-electra-180g-base-discriminator/README.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+谷歌和斯坦福大学发布了一种名为 ELECTRA 的新预训练模型，与 BERT 及其变体相比，该模型具有非常紧凑的模型尺寸和相对具有竞争力的性能。 为进一步加快中文预训练模型的研究，HIT与科大讯飞联合实验室（HFL）发布了基于ELECTRA官方代码的中文ELECTRA模型。 与 BERT 及其变体相比，ELECTRA-small 只需 1/10 的参数就可以在几个 NLP 任务上达到相似甚至更高的分数。
+这个项目依赖于官方ELECTRA代码: https://github.com/google-research/electra
+该模型是base版本的discriminator，并且在180G的中文数据上进行训练。
+
+# 使用示例
+
+```python
+from paddlenlp.transformers import ElectraDiscriminator,ElectraTokenizer
+
+path = "hfl-chinese-electra-180g-base-discriminator"
```
Comment (yingyibiao) [inline community/junnyu/hfl-chinese-electra-180g-base-discriminator/README.md]: path需要进行修改

**Comment by yingyibiao** [comment]
https://github.com/PaddlePaddle/PaddleNLP/blob/develop/docs/model_zoo/transformers.rst
这个文件也需要同步修改

File: paddlenlp/transformers/electra/modeling.py (hunk 1 above)
Comment (JunnYu) [reply paddlenlp/transformers/electra/modeling.py]: 判断不冗余！查看bert的例子，这样会报错！
```python
import paddle
from paddlenlp.transformers import BertModel,BertTokenizer

path = "bert-base-chinese"
model = BertModel.from_pretrained(path)
tokenizer = BertTokenizer.from_pretrained(path)
model.eval()

text = "欢迎使用paddlenlp！"
inputs = {
    k: paddle.tile(paddle.to_tensor(
        v, dtype="int64").unsqueeze(0),[2,1])
    for k, v in tokenizer(text,return_attention_mask=True).items()
}

with paddle.no_grad():
    hidden_states = model(**inputs)[0]

print(hidden_states.shape)
```

Comment (yingyibiao) [reply paddlenlp/transformers/electra/modeling.py]: You are right!

Comment (yingyibiao) [reply paddlenlp/transformers/electra/modeling.py]: 这里我们支持的attention_mask的dtype有int，bool和float。你的修改默认的dtype是int，bool，缺乏对float的支持。

**Review by yingyibiao — COMMENTED** [review]
LGTM

**Review by yingyibiao — APPROVED** [review]
LGTM
